### PR TITLE
feat: bulk sharing API [DHIS2-11787]

### PR DESCRIPTION
**Pull request by vietnguyen**
https://jira.dhis2.org/browse/DHIS2-11787
docs PR: https://github.com/dhis2/dhis2-docs/pull/897

### Using `/api/{object-type}/sharing` with `PATCH` request
- This endpoint allow user to send a Json Patch request to a single object-type which should be used by the maintenance app.
- Note that we still support JSON PATCH request for one object with endpoint `api/{object-type}/{uid}`. Without path variable `uid` the request will be treated as a bulk patch request.
Example: `api/dataElements/sharing`

### Using `/api/metadata/sharing` with `PATCH` request
- We  will also support bulk `PATCH` request for multiple object-types in one payload. This could be a new feature in Metadata Import/Export app.
- This endpoint should be used by the import/export app.

## Parameters

| Name  |  Default  |  Description  |
| ---- | ---- | -------------------- |
| atomic | false | If this is set to true, then the batch function will stop and not updating any objects if there is an error <br> Otherwise, if this is false then the function will try to proceed with best effort mode. |

## Payload
- Payload for single object type using `/api/{object-type}/sharing` looks like this
```json
{
  "dataSets":[
    "cYeuwXTCPkU",
    "aYeuwXTCPkU"
  ],
  "patch":[
    {
      "op":"add",
      "path":"/sharing/users/DXyJmlo9rge",
      "value":{
        "access":"rw------",
        "id":"DXyJmlo9rge"
      }
    },
    {
      "op":"remove",
      "path":"/sharing/users/N3PZBUlN8vq"
    }
  ]
}
```

- Payload for multiple object types in one payload using `api/metadata/sharing`
```json
{
  "dataElements": {
    "fbfJHSPpUQD": [
      {
        "op": "replace",
        "path": "/sharing/users",
        "value": {
          "NOOF56dveaZ": {
            "access": "rw------",
            "id": "NOOF56dveaZ"
          },
          "Kh68cDMwZsg": {
            "access": "rw------",
            "id": "Kh68cDMwZsg"
          }
        }
      }
    ],
    "cYeuwXTCPkU": [
      {
        "op": "replace",
        "path": "/sharing/users",
        "value": {
          "NOOF56dveaZ": {
            "access": "rw------",
            "id": "NOOF56dveaZ"
          },
          "Kh68cDMwZsg": {
            "access": "rw------",
            "id": "Kh68cDMwZsg"
          }
        }
      }
    ]
  },
  "dataSets": {
    "em8Bg4LCr5k": [
      {
        "op": "replace",
        "path": "/sharing/users",
        "value": {
          "NOOF56dveaZ": {
            "access": "rw------",
            "id": "NOOF56dveaZ"
          },
          "Kh68cDMwZsg": {
            "access": "rw------",
            "id": "Kh68cDMwZsg"
          }
        }
      }
    ]
  }
}
```

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchParameters.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2004-2004-2021, University of Oslo
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchParameters.java:2]: Should be `2004-2021` :) several places

**Review by mortenoh — APPROVED** [review]

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    @Test
+    public void testApplyPatchOk()
+        throws IOException
+    {
+        assertStatus( HttpStatus.CREATED, POST( "/users", createUser( "userA", userAId ) ) );
+        assertStatus( HttpStatus.CREATED, POST( "/users", createUser( "userB", userBId ) ) );
+        userCId = assertStatus( HttpStatus.CREATED, POST( "/users", createUser( "userC" ) ) );
+
+        assertStatus( HttpStatus.CREATED,
+            POST( "/dataElements", jsonMapper.writeValueAsString( createDataElement( 'A', deAId, userCId ) ) ) );
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:80]: An alias helper method for `jsonMapper.writeValueAsString`, like `toJsonString`, would keep the overall code more on the semantic level - the mapper itself is quite technical and breaks a bit the otherwise nice readability.

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+        assertStatus( HttpStatus.CREATED,
+            POST( "/dataElements", jsonMapper.writeValueAsString( createDataElement( 'B', deBId, userCId ) ) ) );
+
+        String payload = IOUtils.toString( new ClassPathResource( "patch/bulk_sharing_patch.json" ).getInputStream(),
+            StandardCharsets.UTF_8 );
+        assertStatus( HttpStatus.OK, PATCH( "/dataElements/sharing", payload ) );
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:87]: I think it should just work to use the filename itself as body. no need doing the `getInputStream` part here as it is done here https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/WebClient.java#L226-L232

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            StandardCharsets.UTF_8 );
+
+        HttpResponse response = PATCH( "/metadata/sharing", payload );
+        assertEquals( HttpStatus.CONFLICT, response.status() );
+        assertEquals( "Sharing is not enabled for this object `organisationUnit`", getFirstErrorMessage( response ) );
+    }
+
+    private String getFirstErrorMessage( HttpResponse response )
+    {
+        return response.error().get( "response.typeReports[0].objectReports[0].errorReports[0].message" )
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:208]: FYI: There is also the underlying `JsonNode` tree API which has a `find` method that could be what you want. But if you always want to get the first (and I assume only) error this is the way.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/metadata/MetadataImportExportController.java [hunk 5]
```diff
@@ -125,6 +137,12 @@
     @Autowired
     private ObjectFactory<GmlAsyncImporter> gmlAsyncImporterFactory;
 
+    @Autowired
+    protected ObjectMapper jsonMapper;
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/metadata/MetadataImportExportController.java:141]: I guess we can make both new fields `private`?

**Review by jbee — COMMENTED** [review]
Look good overall. 
I think the patch validators should get an interface and be stored as list so make this more decoupled. The use of `Optional` made the service a bit hard to follow but that is something we can improve in the future. Otherwise some detail comments.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/metadata/MetadataImportExportController.java (hunk 5 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/metadata/MetadataImportExportController.java]: fixed

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java (hunk 4 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:213]: ok noted

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java (hunk 3 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:88]: fixed

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java (hunk 2 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:81]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchParameters.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import lombok.Builder;
+
+import org.hisp.dhis.feedback.ErrorReport;
+
+@Builder
+public class BulkPatchParameters
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchParameters.java]: agree, done. I also replaced the `List<ErrorReport>` with `  private final Map<Class<?>, TypeReport> typeReportMap = new HashMap<>();`  This reduce the code for merging those objects in the controller

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchManager.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    /**
+     * Try to apply given {@link JsonPatch} to given entity by calling
+     * {@link JsonPatchManager#apply}.
+     * <p>
+     * If there is an error, add it to the given errors list and return
+     * {@link Optional#empty()}.
+     */
+    private Optional<IdentifiableObject> applyWithTryCatch( Schema schema, JsonPatch patch, IdentifiableObject entity,
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchManager.java]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchManager.java [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+            if ( !schema.isPresent() )
+            {
+                continue;
+            }
+
+            Map<String, JsonPatch> mapPatches = patches.get( className );
+
+            mapPatches.keySet()
+                .forEach( id -> patchObject( id, schema.get(), mapPatches.get( id ), patchParameters )
+                    .ifPresent( patchedObjects::add ) );
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchManager.java]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchManager.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        if ( !schema.isPresent() )
+        {
+            return Collections.emptyList();
+        }
+
+        List<IdentifiableObject> patchedObjects = new ArrayList<>();
+
+        bulkJsonPatch.getIds()
+            .forEach( id -> patchObject( id, schema.get(), bulkJsonPatch.getPatch(), patchParameters )
+                .ifPresent( patched -> patchedObjects.add( patched ) ) );
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchManager.java]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchManager.java [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+@Service
+public class BulkPatchManager
+{
+    private final IdentifiableObjectManager manager;
+
+    private final SchemaService schemaService;
+
+    private final JsonPatchManager jsonPatchManager;
+
+    public BulkPatchManager( SchemaService schemaService, IdentifiableObjectManager manager,
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchManager.java]: fixed

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ObjectReport.java [hunk 11]
```diff
@@ -111,6 +111,13 @@ public ObjectReport( ObjectReport objectReport )
         this.displayName = objectReport.getDisplayName();
     }
 
+    public ObjectReport( Class<?> klass, int index, List<ErrorReport> errorReports )
+    {
+        this.klass = klass;
+        this.index = index;
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ObjectReport.java]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/patch/BulkPatchManagerTest.java [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+        assertEquals( 1, patchParameters.getErrorReportsCount( ErrorCode.E6002 ) );
+    }
+
+    @Test
+    public void testApplyPatchInvalidUid()
+        throws IOException
+    {
+        final BulkJsonPatch bulkJsonPatch = jsonMapper.readValue(
+            new ClassPathResource( "patch/bulk_sharing_patch_invalid_uid.json" ).getInputStream(),
+            BulkJsonPatch.class );
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/patch/BulkPatchManagerTest.java:170]: Just noticed that these 3 lines are reoccurring in setup and could easily be made more readable with a `loadPatch(filename)` helper method.

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.http.HttpStatus;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @author viet@dhis2.org
+ */
+public class BulkPatchSharingTest extends DhisControllerConvenienceTest
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:55]: All spring MVC controller tests use the naming suffix `ControllerTest`.

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+    private String toJsonString( Object object )
+        throws JsonProcessingException
+    {
+        return jsonMapper.writeValueAsString( object );
+    }
+
+    private String getFirstErrorMessage( HttpResponse response )
+    {
+        return response.error().get( "response.typeReports[0].objectReports[0].errorReports[0].message" )
+            .node().value().toString();
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:214]: Maybe I didn't see it or you got confused by comment about the `JsonNode` API and changed this line using it. If you just want to extract the value of the message the best way is:
```java
response.error().getString( "response.typeReports[0].objectReports[0].errorReports[0].message" ).string();
```

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+        assertStatus( HttpStatus.CREATED, POST( "/users", createUser( "userA", userAId ) ) );
+        assertStatus( HttpStatus.CREATED, POST( "/users", createUser( "userB", userBId ) ) );
+        userCId = assertStatus( HttpStatus.CREATED, POST( "/users", createUser( "userC" ) ) );
+
+        assertStatus( HttpStatus.CREATED,
+            POST( "/dataElements", jsonMapper.writeValueAsString( createDataElement( 'B', deBId, userCId ) ) ) );
+
+        String payload = IOUtils.toString( new ClassPathResource( "patch/bulk_sharing_patch.json" ).getInputStream(),
+            StandardCharsets.UTF_8 );
+        HttpResponse response = PATCH( "/dataElements/sharing?atomic=true", payload );
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:144]: Could also use filename instead of `payload`

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+        assertStatus( HttpStatus.CREATED,
+            POST( "/dataSets", jsonMapper.writeValueAsString( createDataSet( 'A', dsIdA, userCId ) ) ) );
+        assertStatus( HttpStatus.CREATED,
+            POST( "/dataElements", jsonMapper.writeValueAsString( createDataElement( 'A', deAId, userCId ) ) ) );
+        assertStatus( HttpStatus.CREATED,
+            POST( "/dataElements", jsonMapper.writeValueAsString( createDataElement( 'B', deBId, userCId ) ) ) );
+
+        String payload = IOUtils.toString( new ClassPathResource( "patch/bulk_sharing_patches.json" ).getInputStream(),
+            StandardCharsets.UTF_8 );
+        assertStatus( HttpStatus.OK, PATCH( "/metadata/sharing", payload ) );
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:170]: Could also use filename instead of `payload`

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+
+    @Test
+    public void testApplyPatchesInvalidClass()
+        throws IOException
+    {
+        String payload = IOUtils.toString(
+            new ClassPathResource( "patch/bulk_sharing_patches_invalid_class.json" ).getInputStream(),
+            StandardCharsets.UTF_8 );
+
+        HttpResponse response = PATCH( "/metadata/sharing", payload );
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java:200]: Could also use filename instead of `payload`

**Review by jbee — COMMENTED** [review]
Noticed some more details.
Did you plan on creating a common interface for the validators?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchParameters.java (hunk 1 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchParameters.java]: fixed. there are some more files have same issue, i will update other files in another PR

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java (hunk 17 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java]: fixed

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java (hunk 16 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java]: fixed

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java (hunk 15 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java]: fixed

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java (hunk 14 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java]: fixed

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java (hunk 13 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/BulkPatchSharingTest.java]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/patch/BulkPatchManagerTest.java (hunk 12 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/patch/BulkPatchManagerTest.java]: fixed

**Comment by vietnguyen** [comment]
> Noticed some more details.
> Did you plan on creating a common interface for the validators?

Added interface

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchValidators.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+ */
+public interface BulkPatchValidators
+{
+    boolean validateJsonPatch( JsonPatch jsonPatch, Consumer<List<ErrorReport>> errorReportConsumer );
+
+    boolean validateSchema( Schema schema, Consumer<List<ErrorReport>> errorReportConsumer );
+
+    boolean validatePatchEntity( Schema schema, JsonPatch jsonPatch, String id,
+        IdentifiableObject identifiableObject, Consumer<List<ErrorReport>> errorReportConsumer );
+}
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchValidators.java:51]: I see that you changed from returning reports to consumers but I was expecting the other change we talked about changing the validator itself to something like:
```java
interface BulkPatchValidator {
  boolean validate(Schema schema, JsonPatch jsonPatch, Consumer<ErrorReport> addErrors);
}
``` 
As I indicate in the above when using a consumer it makes little sense to use `Consumer<List<T>>` since we changed to consumer to get away from the need to build intermediate objects and collection. Errors can just be added as you go. If there are many you pass them to the consumer right where they are created.

I think if you need the second level validation with the object itself and you still think this should not be done before the update in general is deemed valid it would be OK to have a validator interface with two methods.
```java
interface BulkPatchValidator {
  default boolean validate(Schema schema, JsonPatch jsonPatch, Consumer<ErrorReport> addErrors) {
    return true;
  }
  default boolean validate(Schema schema, JsonPatch jsonPatch, String id, IdentifiableObject identifiableObject, Consumer<ErrorReport> addErrors) {
    return true;
  }
}
``` 
By making them default methods one can override one of them or both when implementing a specific validator.

Then with the above type of validator interface there is no need for a `BulkPatchValidators` class.
The parameters can simply hold a `List<BulkPatchValidator>`. That was the important point so that further additions of validator do not cause changes in further classes. Instead one just uses the existing code passing other data (here validator) to it. Hope that makes it more clear.

Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchValidators.java]: fixed now

**Review by jbee — APPROVED** [review]
Looks good now :+1:
